### PR TITLE
Minor sbt clean-up

### PR DESCRIPTION
**Pull request by james64**
Hey guys,
  I am flirting with contributing to snb. Nothing big yet. I just wanted to get rid of some sbt warnings which I see when I try to explore the code. First I wanted to keep it for myself but then I thought you might like it.

This is mainly about removing deprecated operators (`<<=`, `<+=`, `<++=`) but you can find some other small changes there. Let me know what do you think. I think new code should be doing exactly what old one was.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -296,24 +300,40 @@ lazy val sparkNotebook = project.in(file(".")).enablePlugins(PlayScala).enablePl
...
+      if(withHive.value) "-with-hive" else ""
+    ).mkString,
+
+    version in Debian := Seq(
+      s"${version.in(ThisBuild).value}",
+      s"-scala-${scalaVersion.value}",
+      s"-spark-${sparkVersion.value}",
+      s"-hadoop-${hadoopVersion.value}",
+      if(withHive.value) "-with-hive" else ""
+    ).mkString,
```
Comment (vidma) [inline build.sbt]: should exract this as a helper fn, as it's repeated 3 times...

File: build.sbt [hunk 2]
```diff
@@ -227,12 +229,14 @@ libraryDependencies ++= List(
 )
 
 //for aether and compensating for 2.11 modularization
-libraryDependencies <++= scalaBinaryVersion {
-  case "2.10" => Nil
-  case "2.11" => List(ningAsyncHttpClient,
-    "org.scala-lang.modules" %% "scala-xml" % "1.0.4",
-    "org.scala-lang.modules" %% "scala-parser-combinators" % "1.0.4"
-  )
+libraryDependencies ++= {
```
Comment (vidma) [inline build.sbt]: I see in some places you added extra braces (`{ .... }`) around an expression and in some using the value directly (smaller diff), e.g. https://github.com/spark-notebook/spark-notebook/pull/915/files#diff-fdc3abdfd754eeb24090dbd90aeec2ceR304

do you know if these behave some or not?

**Comment by james64** [comment]
I used [this page](http://www.scala-sbt.org/0.13/docs/Migrating-from-sbt-012x.html) from sbt documentation for this PR.

I am going to manually test those docker etc versions...

File: build.sbt (hunk 1 above)
Comment (james64) [reply build.sbt]: I was thinking about that as well. Will do...

**Comment by james64** [comment]
I compared strings (versions, scripts) generated as can be seen [in a gist here](https://gist.github.com/james64/256ee66deea6f0ab1efd2b40f465b6c7). Only difference I have found is that `common/version` is different. In `master` a value of `common/version` and `spark/version` is different. In my PR it is the same. Have a look in linked gist...

Should I update this so it is consistent with master or maybe this is better?

**Comment by james64** [comment]
Regarding extracting three times repeated code into helper function I was able to came out with two solutions only.

First is to create a separate settingKey for that:
```
lazy val compoundKey = settingKey[String]("")
...
.settings(
    compoundVersion := Seq(
      s"${version.in(ThisBuild).value}",
      s"-scala-${scalaVersion.value}",
      s"-spark-${sparkVersion.value}",
      s"-hadoop-${hadoopVersion.value}",
      if(withHive.value) "-with-hive" else ""
    ).mkString,

    version in Universal := compoundVersion.value,
    version in Docker := compoundVersion.value,
    version in Debian := compoundVersion.value,
)
```

Second is to use function:
```
def f(v: String, scv: String, spv: String, hv: String, wh: String) : String =
  s"${v}-scala-${scv}-spark-${spv}-hadoop-${hv}${wh}"

.settings(
  version in Universal := f(
    version.in(ThisBuild).value,
    scalaVersion.value,
    sparkVersion.value,
    hadoopVersion.value,
    (if(withHive.value) "-with-hive" else ""
  )
// similar for other versions
)
```

It feels weird to create separate global key just to use as tmp variable in one settings in first approach.

Second approach does not really help to remedy repeatability.

**Comment by vidma** [comment]
can't we do something like this (didn't have time to check):

```scala
val fullVersionString = Seq(
      s"${version.in(ThisBuild).value}",
      s"-scala-${scalaVersion.value}",
      s"-spark-${sparkVersion.value}",
      s"-hadoop-${hadoopVersion.value}",
      if(withHive.value) "-with-hive" else ""
    ).mkString


...
something.settings(
  version in Universal := fullVersionString,
    version in Docker := fullVersionString,
    version in Debian := fullVersionString
)
```

**Comment by james64** [comment]
That's exactly how I expected to do it. But sbt says:
```
.../build.sbt:297: error: `value` can only be used within a task or setting macro, such as :=, +=, ++=, Def.task, or Def.setting.
val a = { Seq(s"${scalaVersion.value}ahoj") }
```
So I tried the version with `Def.setting` and that works.

File: build.sbt (hunk 2 above)
Comment (james64) [reply build.sbt]: This relates to usage of `.value` notation. It is not allowed (see sbt error few comments below) to use the notation outside the scope of setting expression. It looks like `.value` is not considered to be within setting expression unless it is enclosed by braces or "hidden" inside string substitution. In such cases I need to use additional braces to make sbt happy. I am not sure if this is a sbt bug or not...

Comment (james64) [reply build.sbt]: I don't think additional braces should change a way how it behaves. I am not dead sure however. I wanted to try to see libraryDependencies content for two versions of scala but as scala 2.10 is not supported anymore (exception raised) I didn't...

File: build.sbt [hunk 3]
```diff
@@ -367,7 +376,7 @@ lazy val observable = Project(id = "observable", base = file("modules/observable
 lazy val common = Project(id = "common", base = file("modules/common"))
   .dependsOn(observable, sbtDependencyManager)
   .settings(
-    version  <<= (version in ThisBuild, sparkVersion) { (v,sv) => s"${sv}_${v}" }
+    version := s"${version.in(ThisBuild).value}_${sparkVersion.value}"
```
Comment (vidma) [inline build.sbt]: it used to be `$sparkVersion_$version`.
need to invert order, for snapshots not to be messed up in Maven repo!!

actually, we spotted a tiny inconsistency with https://github.com/spark-notebook/spark-notebook/pull/915/files#diff-fdc3abdfd754eeb24090dbd90aeec2ceL398 which was inverted..

how about extracting a val here too, e.g. `versionShortWithSpark` ?

see https://gist.github.com/james64/256ee66deea6f0ab1efd2b40f465b6c7#file-versions-comparison-sbt-log-L52

File: build.sbt (hunk 2 above)
Comment (vidma) [reply build.sbt]: so it seems both is valid now as long as you use `.value`
http://www.scala-sbt.org/0.13/docs/Migrating-from-sbt-012x.html#Migrating+from+the+tuple+enrichments

**Comment by james64** [comment]
Maybe we can keep this open few more days. I might try to address following as well through weekend:

```
.../spark-notebook/modules/sbt-dependency-manager/build.sbt:19: warning: object CustomPomParser in package sbt is deprecated: We now use an Aether-based pom parser.
        CustomPomParser.SbtVersionKey -> sbtVersion.reverse.dropWhile(_ != '.').drop(".".length).reverse,
        ^
.../spark-notebook/modules/sbt-dependency-manager/build.sbt:20: warning: object CustomPomParser in package sbt is deprecated: We now use an Aether-based pom parser.
        CustomPomParser.ScalaVersionKey -> scalaBinaryVersion
        ^
```

So we get rid of all warnings when I am on it...

**Comment by vidma** [comment]
if you wish. though, I guess there's more serious issues to fix than warnings ;) just an idea... ;]

**Comment by james64** [comment]
My view was this. By spending an hour researching aether I can learn something which can help me understand the project more. Spending a minute by removing a warning then is a no brainer...

But since I am not able to google up a homepage or anything relevant to aether I am pronouncing this PR done :)

**Comment by james64** [comment]
Can we merge this please?

**Comment by vidma** [comment]
thanks!

**Pull request merged**